### PR TITLE
Delete custom types saga

### PR DESCRIPTION
**Pull request by danigo99**
## Context

Updating state on CT deletion.
Ticket: https://linear.app/prismic/issue/SM-865/[delete-ct]-aauser-i-can-see-the-deleted-custom-types-disappear-from

## The Solution

Added delete CT saga, actions and reducer.
Added a loading state and error toasters on response from the modal.

## Impact / Dependencies

Waiting on https://github.com/prismicio/slice-machine/pull/683

## Checklist before requesting a review
- [ ] I hereby declare my code ready for review.
- [ ] If it is a critical feature, I have added tests.
- [ ] The CI is successful.
- [ ] If there could backward compatibility issues, it has been discussed and planned.

## [OPT] Preview

ezgif com-gif-maker (6)

---

image

### CODE REVIEW

File: packages/slice-machine/src/modules/availableCustomTypes/index.ts [hunk 1]
```diff
@@ -226,6 +256,49 @@ export function* renameCustomTypeSaga({
...
+      );
+    } else {
+      yield put(
+        openToasterCreator({
+          message:
+            "An unexpected error happened while deleting your custom type.",
+          type: ToasterType.ERROR,
+        })
+      );
+    }
```
Comment (danigo99) [inline packages/slice-machine/src/modules/availableCustomTypes/index.ts:288]: This wasn't specified in the ticket, but it's only for the case where the error caught isn't an Axios one. Should never happen but was just to be safe.

File: packages/slice-machine/tests/src/modules/availableCustomTypes/index.test.ts [hunk 2]
```diff
@@ -297,4 +337,83 @@ describe("[Available Custom types module]", () => {
...
+        deleteCustomTypeCreator.request(actionPayload)
+      );
+
+      jest.spyOn(axios, "isAxiosError").mockImplementation(() => true);
+
+      const err = Error() as AxiosError;
+      // @ts-expect-error Ignoring the type error since we only need these properties to test
+      err.response = {
+        data: { reason: "Could not delete custom type", type: "error" },
+      };
```
Comment (danigo99) [inline packages/slice-machine/tests/src/modules/availableCustomTypes/index.test.ts:408]: This isn't great but haven't been able to find a much better alternative

File: packages/slice-machine/src/modules/availableCustomTypes/index.ts [hunk 3]
```diff
@@ -161,6 +182,15 @@ export const availableCustomTypesReducer: Reducer<
         [id]: newCustomType,
       };
     }
+
+    case getType(deleteCustomTypeCreator.success): {
+      const id = action.payload.customTypeId;
+
+      delete state[id];
```
Comment (mathieud) [inline packages/slice-machine/src/modules/availableCustomTypes/index.ts]: The redux state is immutable, don't think you can use `delete` on it. I'm actually surprised it works (unless there's a call to `/state` just after).

You can use

```
return {
  ...state,
  id: undefined
}
``

File: packages/slice-machine/src/modules/availableCustomTypes/index.ts [hunk 4]
```diff
@@ -226,6 +256,49 @@ export function* renameCustomTypeSaga({
...
 }
 
+export function* deleteCustomTypeSaga({
+  payload,
+}: ReturnType<typeof deleteCustomTypeCreator.request>) {
+  try {
+    yield call(deleteCustomType, payload.customTypeId);
+    yield put(deleteCustomTypeCreator.success(payload));
+    yield put(
+      modalCloseCreator({ modalKey: ModalKeysEnum.DELETE_CUSTOM_TYPE })
```
Comment (mathieud) [inline packages/slice-machine/src/modules/availableCustomTypes/index.ts]: This happens in both cases (try and except), you could put it just once at the end

File: packages/slice-machine/src/modules/availableCustomTypes/index.ts [hunk 5]
```diff
@@ -226,6 +256,49 @@ export function* renameCustomTypeSaga({
...
+        type: ToasterType.SUCCESS,
+      })
+    );
+  } catch (e) {
+    yield put(
+      modalCloseCreator({ modalKey: ModalKeysEnum.DELETE_CUSTOM_TYPE })
+    );
+    if (axios.isAxiosError(e)) {
+      const apiResponse = e.response?.data as DeleteCustomTypeResponse;
+      console.log(apiResponse);
```
Comment (mathieud) [inline packages/slice-machine/src/modules/availableCustomTypes/index.ts]: Do we want to keep this?

File: packages/slice-machine/components/DeleteCTModal/index.tsx [hunk 6]
```diff
@@ -104,11 +110,14 @@ export const DeleteCustomTypeModal: React.FunctionComponent<
...
               variant="danger"
-              // eslint-disable-next-line @typescript-eslint/no-misused-promises
-              onClick={async () => {
-                await deleteCustomType(customType?.local.id as string);
-                closeDeleteCustomTypeModal();
-              }}
+              isLoading={isDeletingCustomType}
+              onClick={() =>
+                deleteCustomType(
+                  customType?.local.id ?? "",
```
Comment (mathieud) [inline packages/slice-machine/components/DeleteCTModal/index.tsx]: It should never happen, but do you know what happens if `customType` is actually undefined here?

Comment (danigo99) [reply packages/slice-machine/components/DeleteCTModal/index.tsx]: A lot of bad things 😂 it deletes all the customTypes!

**Review by mathieud — APPROVED** [review]

**Comment by danigo99** [comment]
Updated modal: 
image

**Pull request merged**